client/src/util/stateManager/dict-encoded8_array.ts

```diff
@@ -0,0 +1,23 @@
+/**
+ * This module exports a subclass of Int8Array for
+ * categorical data types. The subclass contains the
+ * `codeMapping` dictionary which maps codes to values,
+ * and a helper function `vat` that returns the value at
+ * a particular index.
+ */
+import { CodeMapping } from "./code_mapping_interfaces";
+
+export class DictEncoded8Array extends Int8Array {
```

Comment:
I'm really struggling with this construct from the standpoint of type system abstractions -- this class is logically an array of categories, which means it is _not_ a specialization (subclass) of an Int8Array.  Put another way, I should be able to treat this object like it is just another Int8Array, which clearly is not true in this case.

Given that the Dataframe.__compileColumn flattens the object into an Array (ie. col().values), why bother defining this type as a sub-class of Int8Array at all?

Put another way, why not just define the class as a simple object with two properties:  `codes: Int8Array` and `codeMapping: CodeMapping`?

Reply:
I initially decided to subclass `Int*Array`s, so that the `.col().asArray()` would behave like an array but also contain the `codeMaping` values. This was before I created a new Dataframe column type and added the codeMapping object to the column itself, so that particular rationale may no longer apply.

The other reason I wanted to sublcass `Int*Array` was to make it easier to interface with the Dataframe constructor, as it seemed easier to pass in some type of array vs passing in an object of `codes` and `codeMapping`. In particular, anywhere that dataframe uses `__columns` internally, I would need to add verbose logic to handle that object. In the current implementation, the only place in dataframe we need to modify outside `__compileColumn` is `_subset`.

Reply:
For example, in `at`, we return:
```typescript
return this.__columns[coff][roff];
```
This works out of the box with a subclass, but will not work with an object. 

Also, a particularly challenging spot for me when trying the object interface was in `mapColumns`: the logic for dealing with a dictionary-encoded object would have to be handled externally by the callback function, which currently just assumes all the columns will be some type of array. This is still the case with `DictInt*Array`s because a naive transformation of the underlaying array buffer will destroy the relationship between `codes and `codeMapping` but this seems like the less bad failure.

Reply:
I understand that it works fine if you think of these as "an array of codes", but they are defined as "an array of labels", in which case `at()` fails to provide the abstraction.

All I'm really pointing out is that these new API break the abstraction that each column is an array of the actual value type (label, float, whatever).  This is witnessed by the fact that `at()` works sometimes, but not always -- you have to first ask the column "are you a dict encoded column?", and then call the correct accessor.

In this new world, I think you have two choices of abstraction:
* the column is still logically an array of values (not codes, but the actual values)
* or, the column is some new more complex beast which can't be treated like an array
Either works, but it would be good to be clear and build the type abstraction to match.  It will help minimize complexity above this layer.